File: streampark-flink/streampark-flink-shims/streampark-flink-shims-base/src/main/scala/org/apache/streampark/flink/core/FlinkStreamingInitializer.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.streampark.flink.core

import org.apache.streampark.common.conf.ConfigConst._
import org.apache.streampark.common.enums.ApiType
import org.apache.streampark.common.enums.ApiType.ApiType
import org.apache.streampark.common.util._

import org.apache.flink.api.java.utils.ParameterTool
import org.apache.flink.configuration.Configuration
import org.apache.flink.streaming.api.environment.{StreamExecutionEnvironment => JavaStreamEnv}
import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment
import org.apache.flink.table.api.TableConfig

import java.io.File

import scala.collection.{mutable, Map}
import scala.collection.JavaConversions._
import scala.util.Try

private[flink] object FlinkStreamingInitializer {

  def initialize(args: Array[String], config: (StreamExecutionEnvironment, ParameterTool) => Unit)
      : (ParameterTool, StreamExecutionEnvironment) = {
    val flinkInitializer = new FlinkStreamingInitializer(args, ApiType.scala)
    flinkInitializer.streamEnvConfFunc = config
    (flinkInitializer.configuration.parameter, flinkInitializer.streamEnv)
  }

  def initialize(args: StreamEnvConfig): (ParameterTool, StreamExecutionEnvironment) = {
    val flinkInitializer = new FlinkStreamingInitializer(args.args, ApiType.java)
    flinkInitializer.javaStreamEnvConfFunc = args.conf
    (flinkInitializer.configuration.parameter, flinkInitializer.streamEnv)
  }
}

private[flink] class FlinkStreamingInitializer(args: Array[String], apiType: ApiType)
  extends Logger {

  var streamEnvConfFunc: (StreamExecutionEnvironment, ParameterTool) => Unit = _

  var tableConfFunc: (TableConfig, ParameterTool) => Unit = _

  var javaStreamEnvConfFunc: StreamEnvConfigFunction = _

  var javaTableEnvConfFunc: TableEnvConfigFunction = _

  implicit private[flink] val parameter: ParameterTool = configuration.parameter

  lazy val streamEnv: StreamExecutionEnvironment = {
    val env = new StreamExecutionEnvironment(
      JavaStreamEnv.getExecutionEnvironment(configuration.envConfig))

    apiType match {
      case ApiType.java if javaStreamEnvConfFunc != null =>
        javaStreamEnvConfFunc.configuration(env.getJavaEnv, configuration.parameter)
      case ApiType.scala if streamEnvConfFunc != null =>
        streamEnvConfFunc(env, configuration.parameter)
      case _ =>
    }
    env.getConfig.setGlobalJobParameters(configuration.parameter)
    env
  }

  lazy val configuration: FlinkConfiguration = initParameter()

  def initParameter(): FlinkConfiguration = {
    val argsMap = ParameterTool.fromArgs(args)
    val configMap = parseConfig(argsMap)
    if (!argsMap.has(KEY_APP_CONF()) && configMap.isEmpty) {
      throw new IllegalArgumentException(
        "[StreamPark] Usage:can't fond config, please set \"--conf $path \" in main arguments")
    }
    val appFlinkConf = extractConfigByPrefix(configMap, KEY_FLINK_PROPERTY_PREFIX)
    val appConf = extractConfigByPrefix(configMap, KEY_APP_PREFIX)
    // config priority: explicitly specified priority > project profiles > system profiles
    val parameter = ParameterTool
      .fromSystemProperties()
      .mergeWith(ParameterTool.fromMap(appFlinkConf))
      .mergeWith(ParameterTool.fromMap(appConf))
      .mergeWith(argsMap)

    val flinkConf: Map[String, String] = {
      parameter.get(KEY_FLINK_CONF(), null) match {
        case flinkConf if flinkConf != null =>
          PropertiesUtils
            .loadFlinkConfYaml(DeflaterUtils.unzipString(flinkConf))
            .filter(_._2.nonEmpty)
        case _ => Map.empty
      }
    }

    val envConfig = Configuration.fromMap(flinkConf + appFlinkConf)
    FlinkConfiguration(parameter, envConfig, null)
  }

  def parseConfig(parameterTool: ParameterTool): Map[String, String] = {
    parameterTool.get(KEY_APP_CONF(), null) match {
      case null | "" =>
        logWarn("Usage:can't fond config, now trying to find from jar")
        val configs =
          Set("application.yml", "application.yaml", "application.conf", "application.properties")
        configs
          .find(
            f => Try(this.getClass.getClassLoader.getResource(f).getPath != null).getOrElse(false))
          .map(
            f => {
              val input = this.getClass.getClassLoader.getResourceAsStream(f)
              val content = FileUtils.readString(input)
              val format = f.split("\\.").last.toLowerCase
              readConfig(format, content)
            })
          .getOrElse(return Map.empty[String, String])
      case config =>
        lazy val content = DeflaterUtils.unzipString(config.drop(7))
        lazy val format = config.split("\\.").last.toLowerCase
        val configAsMap = config match {
          case x if x.startsWith("yaml://") => PropertiesUtils.fromYamlText(content)
          case x if x.startsWith("conf://") => PropertiesUtils.fromHoconText(content)
          case x if x.startsWith("prop://") => PropertiesUtils.fromPropertiesText(content)
          case x if x.startsWith("hdfs://") =>
            // If the configuration file with the hdfs, user will need to copy the hdfs-related configuration files under the resources dir
            val text = HdfsUtils.read(x)
            readConfig(format, text)
          case _ =>
            val configFile = new File(config)
            require(
              configFile.exists(),
              s"[StreamPark] Usage: application config file: $configFile is not found!!!")
            val text = FileUtils.readString(configFile)
            readConfig(format, text)
        }
        configAsMap.filter(_._2.nonEmpty)
    }
  }

  private[this] def readConfig(format: String, text: String): Map[String, String] = {
    format match {
      case "yml" | "yaml" => PropertiesUtils.fromYamlText(text)
      case "conf" => PropertiesUtils.fromHoconText(text)
      case "properties" => PropertiesUtils.fromPropertiesText(text)
      case _ =>
        throw new IllegalArgumentException(
          "[StreamPark] Usage: application config file error,must be [yaml|conf|properties]")
    }
  }

  def extractConfigByPrefix(configMap: Map[String, String], prefix: String): Map[String, String] = {
    val map = mutable.Map[String, String]()
    configMap.foreach(
      x =>
        if (x._1.startsWith(prefix)) {
          map += x._1.drop(prefix.length) -> x._2
        })
    map
  }

}

File: streampark-flink/streampark-flink-shims/streampark-flink-shims-base/src/main/java/org/apache/streampark/flink/core/FlinkStreamingInitializer.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.flink.core;

import org.apache.streampark.common.conf.ConfigKeys;
import org.apache.streampark.common.enums.ApiType;
import org.apache.streampark.common.util.DeflaterUtils;
import org.apache.streampark.common.util.FileUtils;
import org.apache.streampark.common.util.HdfsUtils;
import org.apache.streampark.common.util.PropertiesUtils;
import org.apache.streampark.flink.core.conf.FlinkConfiguration;

import org.apache.flink.api.java.utils.ParameterTool;
import org.apache.flink.configuration.Configuration;
import org.apache.flink.table.api.TableConfig;

import java.io.File;
import java.util.HashMap;
import java.util.Map;

/** Initializes Flink streaming execution environment from CLI args and config files. */
public class FlinkStreamingInitializer extends org.apache.streampark.common.util.LoggerSupport {

    protected final String[] args;
    protected final ApiType apiType;

    scala.Function2<org.apache.flink.streaming.api.scala.StreamExecutionEnvironment, ParameterTool, scala.runtime.BoxedUnit> streamEnvConfFunc;
    scala.Function2<TableConfig, ParameterTool, scala.runtime.BoxedUnit> tableConfFunc;
    StreamEnvConfigFunction javaStreamEnvConfFunc;
    TableEnvConfigFunction javaTableEnvConfFunc;

    private FlinkConfiguration configuration;
    private org.apache.flink.streaming.api.scala.StreamExecutionEnvironment streamEnv;

    protected FlinkStreamingInitializer(String[] args, ApiType apiType) {
        this.args = args;
        this.apiType = apiType;
    }

    ParameterTool parameter() {
        return getConfiguration().parameter();
    }

    FlinkConfiguration getConfiguration() {
        if (configuration == null) {
            configuration = initParameter();
        }
        return configuration;
    }

    org.apache.flink.streaming.api.scala.StreamExecutionEnvironment getStreamEnv() {
        if (streamEnv == null) {
            org.apache.flink.streaming.api.scala.StreamExecutionEnvironment env =
                new org.apache.flink.streaming.api.scala.StreamExecutionEnvironment(
                    org.apache.flink.streaming.api.environment.StreamExecutionEnvironment.getExecutionEnvironment(
                        getConfiguration().envConfig()));
            switch (apiType) {
                case JAVA:
                    if (javaStreamEnvConfFunc != null) {
                        javaStreamEnvConfFunc.configuration(env.getJavaEnv(), getConfiguration().parameter());
                    }
                    break;
                case SCALA:
                    if (streamEnvConfFunc != null) {
                        streamEnvConfFunc.apply(env, getConfiguration().parameter());
                    }
                    break;
                default:
                    break;
            }
            env.getConfig().setGlobalJobParameters(getConfiguration().parameter());
            streamEnv = env;
        }
        return streamEnv;
    }

    FlinkConfiguration initParameter() {
        ParameterTool argsMap = ParameterTool.fromArgs(args);
        String configPath = argsMap.get(ConfigKeys.KEY_APP_CONF(), null);
        if (configPath == null || configPath.isEmpty()) {
            throw new ExceptionInInitializerError(
                "[StreamPark] Usage:can't find config,please set \"--conf $path \" in main arguments");
        }
        Map<String, String> configMap = parseConfig(configPath);
        Map<String, String> properConf =
            extractConfigByPrefix(configMap, ConfigKeys.KEY_FLINK_PROPERTY_PREFIX());
        Map<String, String> appConf = extractConfigByPrefix(configMap, ConfigKeys.KEY_APP_PREFIX());

        ParameterTool parameter =
            ParameterTool.fromSystemProperties()
                .mergeWith(ParameterTool.fromMap(properConf))
                .mergeWith(ParameterTool.fromMap(appConf))
                .mergeWith(argsMap);

        Configuration envConfig = Configuration.fromMap(properConf);
        return new FlinkConfiguration(parameter, envConfig, null);
    }

    Map<String, String> parseConfig(String config) {
        String format = config.contains(".")
            ? config.substring(config.lastIndexOf('.') + 1).toLowerCase()
            : "";
        Map<String, String> map = readConfigContent(config, format);
        Map<String, String> filtered = new HashMap<>();
        map.forEach((key, value) -> {
            if (value != null && !value.isEmpty()) {
                filtered.put(key, value);
            }
        });
        return filtered;
    }

    private Map<String, String> readConfigContent(String config, String format) {
        if (config.startsWith("yaml://")) {
            return readConfigText(format, DeflaterUtils.unzipString(config.substring(7)));
        }
        if (config.startsWith("conf://")) {
            return readConfigText(format, DeflaterUtils.unzipString(config.substring(7)));
        }
        if (config.startsWith("prop://")) {
            return readConfigText(format, DeflaterUtils.unzipString(config.substring(7)));
        }
        if (config.startsWith("hdfs://")) {
            try {
                return readConfigText(format, HdfsUtils.read(config));
            } catch (java.io.IOException e) {
                throw new IllegalStateException("Failed to read HDFS config: " + config, e);
            }
        }
        File configFile = new File(config);
        if (!configFile.exists()) {
            throw new IllegalArgumentException(
                "[StreamPark] Usage: application config file: " + configFile + " is not found!!!");
        }
        try {
            return readConfigText(format, FileUtils.readFile(configFile));
        } catch (java.io.IOException e) {
            throw new IllegalStateException("Failed to read config file: " + configFile, e);
        }
    }

    private Map<String, String> readConfigText(String format, String text) {
        switch (format) {
            case "yml":
            case "yaml":
                return PropertiesUtils.fromYamlText(text);
            case "conf":
                return PropertiesUtils.fromHoconText(text);
            case "properties":
                return PropertiesUtils.fromPropertiesText(text);
            default:
                throw new IllegalArgumentException(
                    "[StreamPark] Usage: application config file error,must be [yaml|conf|properties]");
        }
    }

    Map<String, String> extractConfigByPrefix(Map<String, String> configMap, String prefix) {
        Map<String, String> result = new HashMap<>();
        configMap.forEach(
            (key, value) -> {
                if (key.startsWith(prefix)) {
                    result.put(key.substring(prefix.length()), value);
                }
            });
        return result;
    }
}
